### ISSUE TITLE
Rethink `Lazy` and `foldRight`

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#3226]

**Comment by magnus-madsen** [comment]
@ninaandrup I would like you to look at this next. Let me know if you have questions. I propose to work one namespace at a time and then do the type classes last.

**Comment by ninaandrup** [comment]
> @ninaandrup I would like you to look at this next. Let me know if you have questions. I propose to work one namespace at a time and then do the type classes last.

Sounds good. I'll begin tomorrow :)

**Comment by ninaandrup** [comment]
Things to remeber:

- [ ] Update Foldable instance in `List.flix` to use `List.foldRight`

**Comment by magnus-madsen** [comment]
Based on Nina's experience and reviewing the code I have come to the conclusion that the rewrite of `foldRight` is a bit too difficult to use. Nevertheless, the new `foldRight` is still useful and should be used under certain circumstances. Consequently I would now like the following (which means to keep some of the current work, but also to undo a bit of work):

We should have *two* fold rights:

```
pub def foldRight(f: (a, b) -> b & ef, z: b, l: List[a]): b & ef = ...

pub def foldRightWithCont(f: (a, Unit -> b & ef) -> b & ef, z: b, l: List[a]): b & ef = ...
```

This goes for all collections (`List`, `Set`, `Array` etc.) and also for `Foldable`.

In general, in the standard library, we should refactor back towards *using* the simpler `foldRight` when relevant (this involves a bit of undoing, e.g. changing things like `k()` back to `acc`.)

But we should also incorporate `foldRightWithCont` into `Foldable` and use it to express: 

    Foldable.exists
    Foldable.forall
    Foldable.findLeft
    Foldable.memberOf
    Foldable.isEmpty

    Foldable.head
    Foldable.last

As was the original intention. 

In summary:

- There should be two foldRights. 
- We should also use the simpler one, except for the list I have given above.

@ninaandrup What do you think? Does this make sense?

**Comment by ninaandrup** [comment]
> @ninaandrup What do you think? Does this make sense?

Just one question: Every namespace that has an instance of `Foldable` should then both have a `foldRight` and a `foldRightWithCont`?

**Comment by magnus-madsen** [comment]
Yes, I think so

**Comment by ninaandrup** [comment]
List of namespaces with `foldRight`:

- [ ] `Array.flix`
- [ ] `Chain.flix`
- [ ] `DelayList.flix`
- [ ] `DelayMap.flix`
- [ ] `Iterator.flix`
- [ ] `List.flix`
- [ ] `Map.flix`
- [ ] `MutDeque.flix`
- [ ] `MutList.flix`
- [ ] `MutMap.flix`
- [ ] `MutSet.flix`
- [ ] `Nec.flix`
- [ ] `Nel.flix`
- [ ] `Option.flix`
- [ ] `RedBlackTree.flix`
- [ ] `Reducible.flix`
- [ ] `Result.flix`
- [ ] `Set.flix`

- [ ] Add `foldRightWithCont` to `Foldable`

**Comment by ninaandrup** [comment]
> Does this mean that we only need to:
> 
> express
> 
> ```
> Foldable.exists
> Foldable.forall
> Foldable.findLeft
> Foldable.memberOf
> Foldable.isEmpty
> 
> Foldable.head
> Foldable.last
> ```
> 
> using foldRightWithCont?

Yes, I have already done some of the work in https://github.com/flix/flix/pull/3651

**Comment by magnus-madsen** [comment]
Are we done?

Could you look through the functions in `Foldable` and see if any could be expressed more efficiently with `foldRightCont`?

**Comment by ninaandrup** [comment]
> Are we done?
> 
> Could you look through the functions in `Foldable` and see if any could be expressed more efficiently with `foldRightCont`?

I'll look through `Foldable`, and then we are done :)

**Comment by ninaandrup** [comment]
It does not look like any of the functions would benefit from `foldRightWithCont`.

**Comment by magnus-madsen** [comment]
So we are done? 🎆
